File: cpp/questions/canPartition.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int tn = 0;
        for (const auto& n: nums) {
            tn += n;
        }
        if (tn%2 == 1) {
            return false;
        }
        tn /= 2;
        int cs = 0;
        std::vector<bool> dp(tn+1, false);
        dp[0] = true;
        for (const auto& n: nums) {
            cs += n;
            for (int i=std::min(cs, tn); i>=n; --i) {
                if (dp[i-n]) {
                    dp[i] = true;
                }
            }
            if (dp[tn]) {
                return true;
            }
        }
        return false;
    }
};

File: python/questions/canPartition.py
from typing import List

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        tn = sum(nums)
        if tn % 2:
            return False
        tn //= 2
        dp = [False for _ in range(tn+1)]
        dp[0] = True
        for n in nums:
            for i in range(len(dp)-1, -1, -1):
                pp = i - n
                if pp>=0 and dp[pp]:
                    dp[i] = True
            if dp[-1]:
                return True
        return False
